packages/@lwc/engine-core/src/framework/modules/static-parts.ts

```diff
@@ -105,3 +104,39 @@ export function applyStaticParts(
         applyRefs(part, owner);
     }
 }
+
+/**
+ * Reapplies static parts mounting to newly generated vnodes.
+ *
+ * @param n1 - the previous VStatic vnode
+ * @param n2 - the current VStatic vnode
+ * @param root - the root element
+ * @param renderer - the renderer to use
+ */
+export function patchStaticParts(n1: VStatic, n2: VStatic, root: Element, renderer: RendererAPI) {
+    if (!process.env.IS_BROWSER) {
+        return;
+    }
+
+    const { parts: currParts, owner: currPartsOwner } = n2;
+    if (isUndefined(currParts)) {
+        return;
+    }
+
+    const { parts: prevParts } = n1;
+    if (currParts.length !== prevParts?.length) {
+        // It should not be possible to reach this point as patching only occurs when the vnode is the same.
+        if (process.env.NODE_ENV !== 'production') {
+            throw new Error('Expected static parts to be the same for the same element');
+        }
+        // Recover from patching issue
+        return applyStaticParts(root, n2, renderer, false);
+    }
+
+    for (const [i, part] of currParts.entries()) {
+        // Patch only occurs if the vnode is newly generated, which means the part.elm is always undefined
+        part.elm = prevParts[i].elm;
+        // Refs must be updated after every render due to refVNodes getting reset before every render
+        applyRefs(part, currPartsOwner);
+    }
```

Comment:
We already have the cached static parts on the previous vnode so we can just reassign the values from the previous vnode to the current one.

I thought about keeping the list of static parts to a separate map or data structure but since we need to iterate through them to reapply the refs it seemed unnecessary.

Additionally, we may need to iterate through the parts for the remaining work in #3624, ie updating dynamic attributes, styles, classes, etc.

Leaving it as is for now until the remaining work is completed and then plan to revisit this. 

In a future optimization, we may be able to keep track of the parts that have changed and only update those ones instead of iterating through the entire list of parts.